### DISCUSSION TITLE
issue: subscription not working when setting a value on child component's useEffect

### DISCUSSION THREAD

**Discussion by juan-faccio-swoop** [react-hook-form/react-hook-form#8055] [Q&A]
### Version Number

7.28.0

### Codesandbox/Expo snack

https://codesandbox.io/s/serverless-thunder-0td3b2?file=/src/App.js

### Steps to reproduce

1. On parent component subscribe to a field
2. Create a child component with a combobox, and set a value on `useEffect` hook
3. See repro in Codesandbox

### Expected behaviour

Subscription should work when the Child component set the field value. On the example that I've provided, watch is used to enable/disable a button if there is a value.

It is inside a `useEffect` hook on the codesandbox, but it doesn't work without it neither.

**Note:** it works if `setValue` is pushed to the next tick:

`setTimeout(() => setValue("field", options[0].value), 0)`

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
https://react-hook-form.com/api/usewatch#rules

> useWatch execution order matters, which means if you are update form value before the subscription is in place, then the value update will be ignored.

**Comment by juan-faccio-swoop** [comment]
Thanks for the reply @bluebill1049 , then for this use case (which is more complex on my end), what's the suggestion?

I was aware of the `useWatch` rule, that's why I've tried to push to the next tick with a 0 timeout.

**Comment by bluebill1049** [comment]
> I was aware of the useWatch rule, that's why I've tried to push to the next tick with a 0 timeout.

I would not use timeout, perhaps you want to change the execution order for `useWatch` and `setValue`

**Comment by juan-faccio-swoop** [comment]
Have you check the codesandbox? I've added logs also, the execution order is fine, unless I'm missing something:
* Parent component triggers `subscription`
* Child component triggers `setValue`

**Comment by bluebill1049** [comment]
This is related to the same root cause. `watch` subscription will only get set up after `useEffect` at `useForm` level, which is the reason why we recommend the user to supply a `defaultValues` so the first render for `watch` will return data instead of undefined.

1. watch
2. setValue
3. useEffect
4. watch ready to be subscribed for updated

https://react-hook-form.com/api/useform/watch#rules

we simply couldn't set subscription before the render as it creates side effects with strict mode and this has occurred quite a few issues yet have a good solution, hence any subscription needs to occurred at `useEffect` to be stable.
